### ISSUE TITLE
Ray Serve - Multi Host Support + Autoscaling

### ISSUE THREAD

**Issue by allenwang28** [ray-project/ray#39781]
Met with @richardliaw at Ray Summit regarding multi host serving, summarizing what we discussed. Also Richard asked me to open this as a separate issue and tag @jjyao and @scv119 for feedback and/or help identify any gaps!

# Problem Statement

Our primary use case is multi host serving, e.g. when a replica in our deployment is a TPU pod slice. This pod slice consists of multiple ray worker nodes that have TPU VMs and are provisioned as an atomic unit. 

Primary use case is production-grade serving of large models that do not fit within a single host TPU VM (estimated >65B parameters from [here](https://cloud.google.com/blog/products/compute/how-cloud-tpu-v5e-accelerates-large-scale-ai-inference)).

Note that while this primarily targets TPU pod slices, this would also affect other accelerator types (like Gaudi and GPU VM pods connected by higher bandwidth network) that requires a notion of grouping ray workers together in a deployment. 

## Visualization

To demonstrate this concept, here is a snippet of code of what a sample deployment could look like using today's Ray APIs:

```
@serve.deployment(num_replicas=1, route_prefix="/")
class APIIngress:
    def __init__(self, handle):
        self._handle = handle

    async def __call__(self, request):
        return await self._handle.generate.remote(request)


@serve.deployment(autoscaling_config={
    "min_replicas": 1,
    "max_replicas": 256,
  })
class LLMServer:
    # E.g. represents a v5e-16 and we would want up to 256 replicas of this unit
    def __init__(self):
        # v5e-16 is composed of 4-chip VMs...
        num_hosts = 4
        self._shards = [LLMShard(...) for _ in range(num_hosts)]

    async def generate(self, request):
        response_shards = ray.get([s.generate(request) for s in self._shards])
        # join and postprocess results, e.g. beamsearch etc.
        return processed_result


@ray.remote(resources={"TPU": 4})
class LLMShard:
    # E.g. represents a 4-chip v5e VM host
    def __init__(self):
        import jax
        self._model_shard = # load checkpoint...

    def generate(self, request):
        return self._model_shard.generate(request)
```

As an image:

![ray-multihost-serving-deployment drawio](https://github.com/ray-project/ray/assets/9057208/628f663b-c28e-4f7b-b6ac-1f40f86bb5ba)

## Issues
Autoscaling code snippet above will not work because of two problems:
1) How does Ray know "how" to autoscale?
  - autoscaling currently operates based on resource requests, and would not understand e.g. that we want to scale a unit of v5e-16.
2) How does Ray know where to place LLMShards?
  - e.g. when an LLMServer creates shards, they should land on the same TPU pod slice, but Ray currently cannot tell that unique TPU VMs are or are not part of the same TPU pod slice

And a minor usability concern (3): if our `LLMShard` consisted of multiple functionalities past generate, we might have to duplicate function definitions for each function call. 

# Proposal
A better version would be something like this (same as above, skipping the APIIngress):

## Serve API
```
@serve.deployment(tpu="tpu-v5e-16") # or tpu=("v5e", "4x4") using the accelerator config definition
class LLMServer:
    def __init__(self):
        # e.g. per-shard init logic
        self._model_shard = ... # checkpoint load

    def generate(self, request):
        self._model_shard.generate(request)

    @serve.group_handler
    def group_generate_handler(self, request):
        results = await self.generate(request)
        # join and postprocess results, e.g. beamsearch etc.
```
e.g. we introduce:
1) deployment of a TPU type in which we can specify the exact topology we want a deployment replica to represent
2) a `group_handler` decorator - similar to `@serve.batch` but instead this allows us to define our joining and postprocessing strategy within the same deployment. This will help mitigate issue (3)

## Ray Core - Placement Groups
Serve API could lower the `tpu="tpu-v5e-16"` portion to a placement group. We would need to modify the placement group to accept something like this:
```
bundles = [{"TPU_V5E": 4}] * 4
placement_group(bundles, grouped=True)
```
E.g. the bundles will consist of 4 Ray worker nodes each with 4 v5e chips, and the placement group specifies that they should be "grouped."

This way, the autoscaler can deduce from this placement group spec that this should correspond to a TPU pod of 4 v5e-4s, resolving problem (1).

This leaves issue (2), e.g. how do we ensure that the shards reach the right ray workers? While the placement group will hint to Ray Core that we have to take special consideration about how Raylets schedule this task or actor, we still need Ray to know about the relationship between TPU VM hosts if they are part of the same TPU pod slice. We have to manage this at the resource provisioning level.

## Ray Cluster Launcher
For both the VM cluster launcher and KubeRay, we would need to modify the `ray start` command so we can specify the grouping relationship. For instance, that could look like this:
```
ray start --resources={"TPU": 4} --group_id=$TPU_POD_SLICE_ID
```

Fortunately, TPU VM pod slices are created with a single unique ID naming the pod slice, and we should also be able to specify a unique identifier in the ray start command in KubeRay/GKE.

_Originally posted by @allenwang28 in https://github.com/ray-project/ray/issues/38595#issuecomment-1728599042_

**Comment by richardsliu** [comment]
On the Kuberay side, I propose we make each "worker group" map to each TPU pod slice. Specifically we'll need an extension of the "worker group" concept (call it a "TPU worker group").

The high level idea is as follows:
1. Each TPU worker group corresponds to exactly one TPU pod slice. 
2. The size of each TPU worker group is statically known. This can usually be derived from the TPU topology (e.g. "2x2x4" means 4 TPU hosts in one slice, etc).
3. A TPU worker group cannot be resized once created. It can only be created (scale from 0 to N) or deleted (scale from N to 0).
4. For multi-slice autoscaling, this means we can only scale up or down by whole pod slices at a time. For the Kuberay operator this means the creation or deletion of TPU worker groups, instead of specific replicas.

On the Kuberay operator side, we need to perform a few additional steps when reconciling a worker pod: 
* Pods in a TPU worker group need to be scheduled with affinity. Most likely this will require using a topology key derived from the GKE node pool name. This ensures that all workers in a TPU worker group are scheduled on the same slice.
* The Ray process on each Worker needs a “group_id” marker that identifies workers on the same slice. This marker should also be derived from the GKE node pool name. This is exactly the same as the group id mentioned above. This ensures Ray can schedule actors/tasks on the same slice.
* In addition, TPU worker groups can perform additional initialization by adding the following environment variables:
         * TPU_WORKER_ID: Set to the index of the pod in the worker group. Note that this only works if we are creating new worker groups as a whole - we should never add replicas to existing TPU worker groups. Since we map each TPU worker group to a pod slice of known shape, the precise number of workers for each TPU worker group should be statically known.
         * TPU_WORKER_HOSTNAMES: This is a concatenation of all the statically known worker host names. Again, since we know the number of workers statically, this can be deterministically generated.
         * Correspondingly, each TPU worker group needs its own headless service to ensure that the subdomains can be created.

On the Ray autoscaler side, we need the Kuberay Node Provider to support the following:
* When the application makes an additional resource request, the node provider identifies the corresponding TPU worker group that matches with the request.
* Instead of adding replicas to the same group, the node provider patches the Ray Cluster resource with a new worker group. The new worker group will share the same topology key based on GKE node pool name.
* When scaling down, the node provider needs to know that an entire worker group should be terminated instead of one worker node at a time.

@kevin85421 What do you think?

**Comment by richardliaw** [comment]
In order to support this, ideally we have better support for bundle grouping in placement groups (i.e., all bundles should land on a pod with the same label, regardless of what the label is).

Instead, we currently have the following solution. 

1. On each TPU pod slice, we designate one node in the pod slice to be a "tpu-pod-slice-head". This should be designated either with a custom resource or label (new feature, not totally baked). 
2. On each node of the TPU pod slice, a custom resource with $TPU_POD_ID should be specified.
3. So one node should look like: `ray start --address --resources={"tpu-pod-slice-head-{TPU_TYPE}": 1, $TPU_POD_ID: 4}`, rest of nodes should look like `ray start --address --resources={$TPU_POD_ID: 1}`
4. Then, you make sure your deployment always lands on a head node (via the serve deployment API)
5. The deployment then detects the TPU_POD_ID on that head node, and determines how many TPUs are in the pod-slice (say, X). Then provisions X actors assigned to those TPU_POD_IDs.

```
@serve.deployment(ray_actor_options={"resources": {"tpu-pod-slice-head-TPU_TYPE": 1}}
class LLMServer:
    def __init__(self):
        pod_id = os.environ.get("TPU_POD_ID")
        num_tpus_in_pod = calculate_num_tpus_in_pod()
        TPUActor = Actor.options(resources={pod_id: 1, "tpu": 1})
        self._actor_group = [TPUActor.remote() for _ in range(num_tpus_in_pod)]

    def generate(self, request):
        return [actor.generate(request) for actor in self._actor_group]

    def group_generate_handler(self, request):
        results = await self.generate(request)
        # join and postprocess results, e.g. beamsearch etc.
```

**Comment by kevin85421** [comment]
The approach suggested in [this comment](https://github.com/ray-project/ray/issues/39781#issuecomment-1736479133) seems to have a high level of complexity and could lead to alterations in existing behaviors of KubeRay, such as autoscaling, worker group, and new headless services, potentially impacting maintainability. Hence, I have the following proposal:

* GKE
  * Users should create a TPU node pool for each RayCluster, and each node pool has unique taints/labels for Pod scheduling.

* TPU
  * TPU team should provide two utility functions:
    * `get_podslice_id`: Get the unique ID of a PodSlice that this TPU Pod belongs to.
    * `get_tpu_worker_id`: Get a unique integer for each TPU Pod in the same PodSlice.

* KubeRay
  *  KubeRay does not track which TPU Pods belong to which TPU PodSlice. That is, KubeRay will not set `group_id`.
  * Users configure the TPU Pod's nodeSelector and tolerations to ensure the Pod is scheduled on a TPU PodSlice belonging to a specific TPU node pool.
  * KubeRay always respects the decisions made by Ray Autoscaler to create or delete specific Pods. It should not determine whether the Ray Pod is a TPU Pod, nor whether these four Ray Pods belong to the same TPU PodSlice. Ray Autoscaler should be the single source of the truth.

* Ray Core
  * Setup environments
    * `group_id`: Use `get_podslice_id` to set it.
    * `TPU_WORKER_ID`: Use `get_tpu_worker_id` to set it.
    * `TPU_WORKER_HOSTNAMES`: Ray Core, aware of the IP information of each Ray node and the `group_id` information, can handle `TPU_WORKER_HOSTNAMES` gracefully.
  * Scheduling
    * https://github.com/ray-project/ray/issues/39781#issuecomment-1741313233
  * Autoscaling

**Comment by allenwang28** [comment]
https://github.com/ray-project/ray/pull/40463 should enable pod autoscaling for the GCE VM based approach using a similar API laid out in https://github.com/ray-project/ray/issues/39781#issuecomment-1741313233

The Kuberay piece is equally important here though and I am not sure I fully understand if there was a conclusion from https://github.com/ray-project/ray/issues/39781#issuecomment-1736479133 and https://github.com/ray-project/ray/issues/39781#issuecomment-1743658075.

A piece of confusion from my end is that I don't this will work as we expect it to:
> Users should create a TPU node pool for each RayCluster, and each node pool has unique taints/labels for Pod scheduling.

To clarify, a TPU node pool can only schedule a single TPU pod (see [ref](https://cloud.google.com/kubernetes-engine/docs/concepts/tpus#machine_type)) so we can't get our autoscaling from the node pool. That's why @richardsliu suggested the approach in https://github.com/ray-project/ray/issues/39781#issuecomment-1736479133 e.g. scale by nodepools rather than nodes within a nodepool.

**Comment by kevin85421** [comment]
> To clarify, a TPU node pool can only schedule a single TPU pod (see [ref](https://cloud.google.com/kubernetes-engine/docs/concepts/tpus#machine_type)) so we can't get our autoscaling from the node pool. That's why @richardsliu suggested the approach in https://github.com/ray-project/ray/issues/39781#issuecomment-1736479133 e.g. scale by nodepools rather than nodes within a nodepool.

Thanks, @allenwang28! I lost the message when I attended KubeCon. Thank @richardsliu for the reminder. I have several questions:

* Who should scale up the node pools?
* How can we ensure that all TPU hosts in the PodSlice are assigned to the same RayCluster?

**Comment by richardsliu** [comment]
> * Who should scale up the node pools?

In GKE today, each TPU pod slice is mapped to one node pool. So the node pool can only be resized from 0 to x to 0.

> * How can we ensure that all TPU hosts in the PodSlice are assigned to the same RayCluster?

When the Ray Cluster is created, it should create X workers where X is the number of TPU hosts in a PodSlice. We also should add node affinity labels to ensure that all workers are co-scheduled on the same slice. @ryanaoleary is working on a TPU webhook which can validate this.

**Comment by allenwang28** [comment]
> In GKE today, each TPU pod slice is mapped to one node pool. So the node pool can only be resized from 0 to x to 0.

Yeah this piece is still unclear to me - we would want it to be such that we could resize from 0 to x to 2x to 3x, etc. back to 0. But I am not sure if the change should be within KubeRay (i.e. through KubeRay, explore if it's possible scale up multiple NodePools) or through GKE.

**Comment by richardsliu** [comment]
> > In GKE today, each TPU pod slice is mapped to one node pool. So the node pool can only be resized from 0 to x to 0.
> 
> Yeah this piece is still unclear to me - we would want it to be such that we could resize from 0 to x to 2x to 3x, etc. back to 0. But I am not sure if the change should be within KubeRay (i.e. through KubeRay, explore if it's possible scale up multiple NodePools) or through GKE.

Yeah this is the reason for my proposal above - each worker group would map to one GKE node pool, so autoscaling from 0 to x to 2x would end up as creating new worker groups in Kuberay. As far as I know, a TPU node pool in GKE cannot be resized from x to 2x.

**Comment by kevin85421** [comment]
> In GKE today, each TPU pod slice is mapped to one node pool. So the node pool can only be resized from 0 to x to 0.

Sorry for the lack of clarity. I knew the atomicity for the multi-host TPU node pool from the [GKE doc](https://cloud.google.com/kubernetes-engine/docs/concepts/tpus#machine_type) referenced in https://github.com/ray-project/ray/issues/39781#issuecomment-1800328451. My question is: Who should create a new TPU node pool when KubeRay attempts to create new TPU Pods?

> @ryanaoleary is working on a TPU webhook which can validate this.

Do you mean that validating that all Pods in a multi-host TPU node pool should be in the same RayCluster?

**Comment by richardsliu** [comment]
> > In GKE today, each TPU pod slice is mapped to one node pool. So the node pool can only be resized from 0 to x to 0.
> 
> Sorry for the lack of clarity. I knew the atomicity for the multi-host TPU node pool from the [GKE doc](https://cloud.google.com/kubernetes-engine/docs/concepts/tpus#machine_type) referenced in [#39781 (comment)](https://github.com/ray-project/ray/issues/39781#issuecomment-1800328451). My question is: Who should create a new TPU node pool when KubeRay attempts to create new TPU Pods?

This can be done from GKE Node Auto-Provisioning. The user should not need to manually create the node pools.

> 
> > @ryanaoleary is working on a TPU webhook which can validate this.
> 
> Do you mean that validating that all Pods in a multi-host TPU node pool should be in the same RayCluster?

Validating that when a RayCluster reserves worker pods in a TPU pod slice, the number of workers should match the number of hosts in the slice. Also the pods should be co-scheduled with node affinity such that all workers get placed on the same slice. This should guarantee that all pods in the pod slice are assigned to the same Ray Cluster.

**Comment by kevin85421** [comment]
Understood. I think this feature should be implemented in the Ray Autoscaler rather than in KubeRay. If the Ray Autoscaler decides to scale up a multi-host TPU PodSlice, it should update the RayCluster CR to add a new worker group. Then, KubeRay will create new Pods based on the updated CR specification.

**Comment by kevin85421** [comment]
* GKE
  - [ ] GKE Node Auto-Provisioning is responsible for creating a new node pool for each new multi-host TPU PodSlice(https://github.com/ray-project/ray/issues/39781#issuecomment-1823241432).
  - [ ] Provide a TPU webhook to ensure all Pods on the same TPU PodSlice belong to the same RayCluster custom resource (https://github.com/ray-project/ray/issues/39781#issuecomment-1823241432).

* TPU
  * TPU team should provide two utility functions:
    - [x] `get_podslice_id`: Get the unique ID of a PodSlice that this TPU Pod belongs to. I chatted with @jjyao yesterday. I am not sure how it is implemented, but Ray can currently set a unique ID for different TPU PodSlices, so I marked it as done.
    - [ ] `get_tpu_worker_id`: Get a unique integer for each TPU Pod in the same PodSlice. This is used to set the environment variable `TPU_WORKER_ID`. If Ray offers an alternative method to set the environment, we can consider this task as completed.

* KubeRay
  * Scheduling
    * The webhook provided by GKE should promise all Pods on the same TPU PodSlice belong to the same RayCluster custom resource. See the section "GKE" above.
    - [ ] KubeRay may need to add some configurations to enable the webhook to schedule all Pods in a RayCluster TPU worker group should be scheduled to the same TPU PodSlice.
  * Autoscaling
    * Ray Autoscaler should add/delete TPU worker groups to scale up/down TPU PodSlices.
    * KubeRay always respects the decisions made by Ray Autoscaler to create or delete specific Pods. It should not determine whether the Ray Pod is a TPU Pod, nor whether these four Ray Pods belong to the same TPU PodSlice. Ray Autoscaler should be the single source of the truth. Hence, the change should be in Ray Autoscaler rather than KubeRay.

* Ray Core
  - [ ] Setup environments
    - [ ] `TPU_WORKER_ID`: See [Deploy TPU workloads in GKE](https://cloud.google.com/kubernetes-engine/docs/how-to/tpus) for more details.
    - [ ] `TPU_WORKER_HOSTNAMES`: See [Deploy TPU workloads in GKE](https://cloud.google.com/kubernetes-engine/docs/how-to/tpus) for more details.
  - [ ] Scheduling
  - [ ] Autoscaling
    - [ ] VM
    - [ ] KubeRay

cc @jjyao @richardsliu @allenwang28 @ryanaoleary @richardliaw Does this summary make sense to you? Could you also check the status of each item? I think some items may have already been completed. Thanks!
